### PR TITLE
feat(DqElement): Truncate large `html` strings when the element has a large outerHTML string

### PR DESCRIPTION
**Pull request by GeneratorX16**
feat(DqElement): Fixed truncation logic for large outerHTML for the the element itself

Changed truncation logic to also truncate the element of DqElement's source, while preserving the earlier truncation logic.

Closes: Issue #4544

### CODE REVIEW

File: test/core/utils/dq-element.js
Comment (stephenmathieson) [inline test/core/utils/dq-element.js:1]: Can you please add a test for excessively long attribute names? Something like:

```js
const el = document.createElement('div')
let attributeName = 'data-'
for (let i = 0; i < 10000000; i++) {
  attributeName += 'foo'
}
el.setAttribute(attributeName, 'bar')
```

I don't know what the expected behavior will be here, but I'm sure this is a case we'll want to handle.

Comment (GeneratorX16) [reply test/core/utils/dq-element.js:1]: Sure. For excessively long attribute names, it will truncate it to the max attribute length, which here would be 20 characters. So the result for this would be 
`<div data-foofoofoofoofoo...  ></div>`
Is this acceptable ?

Comment (stephenmathieson) [reply test/core/utils/dq-element.js:1]: That makes sense to me 👍

I'll bring this up internally and see what the rest of the team thinks.

Comment (WilcoFiers) [reply test/core/utils/dq-element.js:1]: Looks good to me. @straker can you check to make sure a change like this doesn't break anyone's deduplication code? Otherwise this may need to go behind an option.

File: lib/core/utils/dq-element.js [hunk 1]
```diff
@@ -8,26 +8,85 @@ import memoize from './memoize';
...
+    return '';
+  }
+
+  let str = getOuterHtml(element);
+  if (!truncatedAttributeSet || truncatedAttributeSet.size === 0) {
+    return str;
+  }
+
+  let intermediateAttrString = '';
+  const openingTagEndIndex = str.indexOf('>');
```
Comment (straker) [inline lib/core/utils/dq-element.js]: This makes an assumption that the first use of the `>` character is the end of the tag, but that isn't going to be true as it could be used in an attribute.

File: lib/core/utils/dq-element.js [hunk 2]
```diff
@@ -8,26 +8,85 @@ import memoize from './memoize';
...
+
+  const copyNode = element.cloneNode(true);
+  const attributesToTruncate = new Set();
+
+  const elementAttributes = element.attributes;
+  const lastIndex = elementAttributes.length - 1;
+  let currElementStr = truncateAttributes(copyNode, attributesToTruncate);
+
+  for (let i = lastIndex; i >= 0; i--) {
+    if (currElementStr.indexOf('>') < maxLength) {
```
Comment (straker) [inline lib/core/utils/dq-element.js]: `>` could be used in an attribute and not the closing of the opening tag.

File: lib/core/utils/dq-element.js [hunk 3]
```diff
@@ -8,26 +8,85 @@ import memoize from './memoize';
...
+    str.substring(0, openingTagEndIndex) +
+    intermediateAttrString +
+    ' ' +
+    str.substring(openingTagEndIndex);
   return str;
 }
 
+function truncateElement(element, maxLength, maxAttributeLength) {
+  maxLength = maxLength || 100;
+  maxAttributeLength = maxAttributeLength || 20;
```
Comment (straker) [inline lib/core/utils/dq-element.js]: This function isn't exported so only this file will have access to it. Since we don't change the max options anywhere, could remove them as arguments and instead create them as constants.

File: lib/core/utils/dq-element.js [hunk 4]
```diff
@@ -8,26 +8,85 @@ import memoize from './memoize';
...
+
+  const copyNode = element.cloneNode(true);
+  const attributesToTruncate = new Set();
+
+  const elementAttributes = element.attributes;
```
Comment (straker) [inline lib/core/utils/dq-element.js]: This should have thrown an eslint error warning of not using `element.attributes`. This can cause problems if the node has been clobbered, so should use [getNodeAttributes](https://github.com/dequelabs/axe-core/blob/develop/lib/core/utils/get-node-attributes.js) instead.

I'll have to update our eslint config to look for `element` and not just `node`.

File: lib/core/utils/dq-element.js [hunk 5]
```diff
@@ -8,26 +8,85 @@ import memoize from './memoize';
...
+  const lastIndex = elementAttributes.length - 1;
+  let currElementStr = truncateAttributes(copyNode, attributesToTruncate);
```
Comment (straker) [inline lib/core/utils/dq-element.js]: Why call `truncateAttributes` here if the `attributesToTruncate` is an empty set? Won't that just give you back the outer html of the element?

File: lib/core/utils/dq-element.js [hunk 6]
```diff
@@ -8,26 +8,85 @@ import memoize from './memoize';
...
+    if (currElementStr.indexOf('>') < maxLength) {
+      break;
+    }
+
+    const attribute = elementAttributes[i];
+    const attributeLength = attribute.value.length + attribute.name.length;
+
+    if (attributeLength > maxAttributeLength) {
+      attributesToTruncate.add(attribute.name);
+      copyNode.attributes.removeNamedItem(attribute.name);
```
Comment (straker) [inline lib/core/utils/dq-element.js]: This will be incompatible with the `getNodeAttributes` function as a clobbered element returns a copy of the attributes and not the attributes attached to the DOM node. The code will need to avoid using properties and functions off of a nodes `attributes`.

File: lib/core/utils/dq-element.js [hunk 7]
```diff
@@ -8,26 +8,85 @@ import memoize from './memoize';
...
+      currElementStr = truncateAttributes(copyNode, attributesToTruncate);
```
Comment (straker) [inline lib/core/utils/dq-element.js]: I worry that repeatedly calling `truncateAttributes`, specifically getting the elements outer HTML repeatedly, will be slow for large elements with many DOM nodes (think of doing this on `body` for a page with 100,000+ nodes when body has a long list of attributes). 

I think it would be faster if we either isolated the current element by doing `cloneNode(false)` when working on truncating the attributes, or possibly keeping a list of attributes as a string and using that to build up the element string instead.

File: lib/core/utils/dq-element.js [hunk 8]
```diff
@@ -8,26 +8,85 @@ import memoize from './memoize';
...
+    }
+  }
+
+  if (currElementStr.indexOf('>') > maxLength) {
```
Comment (straker) [inline lib/core/utils/dq-element.js]: `>` could be used in an attribute and not the closing of the opening tag.

**Review by straker — CHANGES_REQUESTED** [review]
Thanks for the pr! I'll go confirm with other teams about their duplication strategy and make sure `source` isn't one of the properties they use. Overall the code and is good. Below I've left some comments and suggestions, mostly focused around using `element.attributes` clobbering and the use of `>` to detect the end of the opening tag.

File: lib/core/utils/dq-element.js (hunk 7 above)
Comment (GeneratorX16) [reply lib/core/utils/dq-element.js]: Yes. This gave me the idea to treat the innerHTML content and the node itself (using shallow node copy)  separately and then handle them and show the result by "merging" them. Is this a better approach ? 

If we do the string method, is it always the case that attribute values will be enclosed in double quotes (") for all elements ? (ex. `<div class="abcd"> ` ? 

Also, would it be better to remove all attributes that are greater than the given length in one go ? Or should I check the string after each truncation to keep it from truncating long attributes even if the string is within the given bounds (current behaviour) ?

File: lib/core/utils/dq-element.js (hunk 1 above)
Comment (GeneratorX16) [reply lib/core/utils/dq-element.js:32]: Would it be better to have a literal of void elements to check if it is a void element or a container element ? Or will calculating it be better (through logic, not through the tag name)

File: lib/core/utils/dq-element.js (hunk 7 above)
Comment (straker) [reply lib/core/utils/dq-element.js]: For the array method you wouldn't want to parse the outer html of the string itself, but instead use the `getNodeAttributes` and work off of that. Something like:

```js
let str = node.outerHTML(false);

// exit early if length is fine
if (str.length < maxLength) {
  return str;  // I can't remember if this includes children or not, so may need to do a cloneNode(true) here instead
}

const attributes = [];
// first pass to reduce size of any long attributes
for ({ name, value } of getNodeAttributes(node)) {
  const attr = { name, value }  // break it from being a live value which would modify the DOM value
  if (attribute.value > maxAttributeLength) {
    attr.value = '...'
  }

  attributes.push(attr);
}

str = '<' + node.nodeName.toLowerCase() + attributes.map({ name, value } => `${name}="${value}"`).join(' ') + '>'
while (str.length > maxLength) {
  attributes.pop();
  str = '<' + node.nodeName.toLowerCase() + attributes.map({ name, value } => `${name}="${value}"`).join(' ') + '...>'
}

return str;  // again, might need to make sure the return type is correct here
```

Comment (GeneratorX16) [reply lib/core/utils/dq-element.js]: I have taken notes from this to build the latest method. It also includes for truncation of long attribute name and long tag name, along with the attribute value too.


Also, made the entire logic have a run time of linear order, so large elements would not considerably slow down the truncation logic.

File: lib/core/utils/dq-element.js (hunk 8 above)
Comment (GeneratorX16) [reply lib/core/utils/dq-element.js]: It is not needed now, as we are constructing the outerHTML string ourselves.

File: lib/core/utils/dq-element.js (hunk 1 above)
Comment (GeneratorX16) [reply lib/core/utils/dq-element.js]: It is not needed now, as we are constructing the outerHTML string ourselves.

File: lib/core/utils/dq-element.js (hunk 2 above)
Comment (GeneratorX16) [reply lib/core/utils/dq-element.js]: It is not needed now, as we are constructing the outerHTML string ourselves.

File: lib/core/utils/dq-element.js (hunk 4 above)
Comment (GeneratorX16) [reply lib/core/utils/dq-element.js]: I have replaced the attributes property with the result of getNodeAttributes.

File: lib/core/utils/dq-element.js (hunk 5 above)
Comment (GeneratorX16) [reply lib/core/utils/dq-element.js]: It is not needed now, as we are constructing the outerHTML string ourselves.

File: lib/core/utils/dq-element.js (hunk 6 above)
Comment (GeneratorX16) [reply lib/core/utils/dq-element.js]: It is not needed now, as we are constructing the outerHTML string ourselves.

**Comment by straker** [comment]
Thanks for the changes. I'm about to leave for a trip so I won't be able to look at this till next week.

**Comment by straker** [comment]
Thanks for the changes. I'm unable to look at this this week but will try to look at it next week.

File: lib/core/utils/dq-element.js [hunk 9]
```diff
@@ -5,29 +5,82 @@ import getNodeFromTree from './get-node-from-tree';
...
+    str = `<${shallowNode.nodeName.toLowerCase()} ${attributeStrList.join(' ')} ...>`;
+  } else {
+    str = `<${shallowNode.nodeName.toLowerCase()} ${attributeStrList.join(' ')}>`;
+  }
+
+  if (str.length > maxLen) {
+    str = str.substring(0, maxLen) + '... >';
   }
 
   return str;
```
Comment (straker) [inline lib/core/utils/dq-element.js:91]: This brings up an interesting question. What should happen if the element itself has too many attributes that it becomes too long, but the children length is fine? Currently something like `<button aria-label="Wednesday the 14th">Wed the 14th</button>` would be fine and show the full outerHTML for the `source`, but if we add a single attribute that makes the string >300, after truncating the string down we would end up with a length that would allow the children and end tag to fit with the 300 max length. @dbjorge @WilcoFiers  Should we try to add the children back in that case or just leave them out?

Example, should the following HTML

```html
<button aria-label="Wednesday the 14th" data-test="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam interdum lorem nisi, quis volutpat dolor efficitur sed. In sollicitudin nibh at nibh viverra, et scelerisque elit sodales. Quisque pellentesque laoreet turpis ac porttitor. Quisque a tristique felis. Curabitur enim neque, accumsan vel malesuada ac, maximus vel lectus. Nam auctor felis ac ullamcorper lobortis. Ut accumsan nulla et mauris varius posuere. Suspendisse placerat sapien metus, nec venenatis arcu porttitor quis. Aenean mollis venenatis velit sit amet bibendum. Nulla porta neque augue, eu condimentum orci malesuada vitae. Aliquam iaculis, felis ac fermentum mattis, tellus orci vehicula dui, malesuada fermentum enim mi sed enim. Donec tempor faucibus accumsan. Nunc id erat purus. Donec eget ipsum accumsan, euismod mauris vel, lacinia nisi.">Wed the 14th</button>
```

result in

```html
<button aria-label="Wednesday the 14th" data-test="Lorem ipsum dolor si...">
```

or

```html
<button aria-label="Wednesday the 14th" data-test="Lorem ipsum dolor si...">Wed the 14th</button>
```

I think personally I'm fine with the first option as it's the simplest.

File: lib/core/utils/dq-element.js [hunk 10]
```diff
@@ -5,29 +5,82 @@ import getNodeFromTree from './get-node-from-tree';
...
+        ? attrValue.substring(0, maxAttrNameOrValueLen) + '...'
+        : attrValue;
+
+    const strAttr = `${attrName}="${attrValue}"`;
+    strLen += (' ' + strAttr).length;
+    attributeStrList.push(strAttr);
+  }
+
+  if (attributeStrList.length < elementNodeMap.length) {
+    str = `<${shallowNode.nodeName.toLowerCase()} ${attributeStrList.join(' ')} ...>`;
```
Comment (straker) [inline lib/core/utils/dq-element.js]: If the document is in XHTML we shouldn't lowercase the node name (see [virtual node class](https://github.com/dequelabs/axe-core/blob/develop/lib/core/base/virtual-node/virtual-node.js#L58)).

File: lib/core/utils/dq-element.js [hunk 11]
```diff
@@ -5,29 +5,82 @@ import getNodeFromTree from './get-node-from-tree';
...
+  } else {
+    str = `<${shallowNode.nodeName.toLowerCase()} ${attributeStrList.join(' ')}>`;
+  }
+
+  if (str.length > maxLen) {
+    str = str.substring(0, maxLen) + '... >';
```
Comment (straker) [inline lib/core/utils/dq-element.js]: This leaves an odd edge case where if there are less `attributeStr` attributes than `elementNodeMap` attributes, you end the string with `...`. However if that final string is then longer than the `maxLen` you also add 3 more dots. So if the first string with the 3 dots is say 2 characters over the `maxLen`, you add more dots and end up with more than 3 dots at the end.

E.g.

```html
<button aria-label="Wednesday the 14th" data-test="Lorem ipsum dolor sit amet, consectetur adipiscing elit." data-test1="Lorem ipsum dolor sit amet, consectetur adipiscing elit." data-test2="Lorem ipsum dolor sit amet, consectetur adipiscing elit." data-test3="Lorem ipsum dolor sit amet, consectetur adipiscing elit." data-test4="Lorem ipsum dolor sit amet, consectetur adipiscing elit." data-test5="Lorem ipsum dolor sit amet, consectetur adipiscing elit." data-test678910="Lorem ipsum dolor sit amet, consectetur adipiscing elit." data-test7="Lorem ipsum dolor sit amet, consectetur adipiscing elit.">Wed the 14th</button>
```

```html
<button aria-label=\"Wednesday the 14th\" data-test=\"Lorem ipsum dolor si...\" data-test1=\"Lorem ipsum dolor si...\" data-test2=\"Lorem ipsum dolor si...\" data-test3=\"Lorem ipsum dolor si...\" data-test4=\"Lorem ipsum dolor si...\" data-test5=\"Lorem ipsum dolor si...\" data-test678910=\"Lorem ipsum dolor si..... >
```

File: lib/core/utils/dq-element.js [hunk 12]
```diff
@@ -5,29 +5,82 @@ import getNodeFromTree from './get-node-from-tree';
...
+ * @param  {Node}   element the element node which needs to be truncated
+ */
+
+export function truncateElement(element) {
+  const maxLen = 300;
+  const maxAttrNameOrValueLen = 20;
+
+  const deepStr = getOuterHtml(element);
+
+  if (deepStr.length < maxLen) {
```
Comment (straker) [inline lib/core/utils/dq-element.js:39]: This code assumes that the reason the outerHTML was long was because of the attributes, and not due to the children. If the children were longer than the max the attributes should probably be left unchanged if they fit with the 300 length. @dbjorge @WilcoFiers what do you think?

E.g.

```html
<button aria-label="Wednesday the 14th" data-test="Lorem ipsum dolor sit amet, consectetur adipiscing elit.">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer et ante ligula. Suspendisse blandit, quam tincidunt hendrerit pulvinar, nisi lacus rutrum nibh, condimentum bibendum ipsum mi vitae neque. Morbi ac interdum lacus. Vivamus id quam in tellus dignissim convallis. Duis ultricies, eros quis aliquet cursus, lectus enim tempor tellus, eget fermentum ante augue sit amet lectus. Aliquam vehicula cursus molestie. Etiam ultricies sapien felis, vitae lacinia nisl sagittis quis.</button>
```

Should probably result in 

```html
<button aria-label="Wednesday the 14th" data-test="Lorem ipsum dolor sit amet, consectetur adipiscing elit.">
```

and not

```html
<button aria-label="Wednesday the 14th" data-test="Lorem ipsum dolor si...">
````

**Review by straker — DISMISSED** [review]
Thanks for the changes. Seeing the changes had me think about a few edge cases that we should make sure we all agree on so pining my colleagues about it.

File: lib/core/utils/dq-element.js (hunk 9 above)
Comment (WilcoFiers) [reply lib/core/utils/dq-element.js:91]: I'd lean towards leaving children off, so that your "..." is near the end of the snippet. That seems like it'd be a clearer truncation. I can go either way on that though.

File: lib/core/utils/dq-element.js (hunk 11 above)
Comment (GeneratorX16) [reply lib/core/utils/dq-element.js:72]: I will add an extra whitespace for the final truncation.
So if it is longer, it will show that the attributes are also truncated, and attributes have been removed too.
```html
<button aria-label=\"Wednesday the 14th\" data-test=\"Lorem ipsum dolor si...\" data-test1=\"Lorem ipsum dolor si...\" data-test2=\"Lorem ipsum dolor si...\" data-test3=\"Lorem ipsum dolor si...\" data-test4=\"Lorem ipsum dolor si...\" data-test5=\"Lorem ipsum dolor si...\" data-test678910=\"Lorem ipsum dolor si..... >
``` 

will be shown as:

```html
`<button aria-label=\"Wednesday the 14th\" data-test=\"Lorem ipsum dolor si...\" data-test1=\"Lorem ipsum dolor si...\" data-test2=\"Lorem ipsum dolor si...\" data-test3=\"Lorem ipsum dolor si...\" data-test4=\"Lorem ipsum dolor si...\" data-test5=\"Lorem ipsum dolor si...\" data-test678910=\"Lorem ipsum dolor si... ... >`
```

File: lib/core/utils/dq-element.js (hunk 12 above)
Comment (GeneratorX16) [reply lib/core/utils/dq-element.js:39]: I have handled this case now. If element is within bounds but the children are causing it to be larger, I do not truncate the element , but discard the children (resulting in the result expected in the comment)

Comment (GeneratorX16) [reply lib/core/utils/dq-element.js:39]: Also, utilized Virtual Node in order to be compatible with Xhtml

File: lib/core/utils/dq-element.js (hunk 9 above)
Comment (GeneratorX16) [reply lib/core/utils/dq-element.js:91]: I have left the children off in this case.

**Comment by straker** [comment]
Thanks for the changes. Unfortunately something in those changes broke a bunch of the tests. Double unfortunately since you forced pushed I'm unable to see the newest changes compared to prior changes, so I don't know exactly what you changed that caused the problem, nor can I revert to the point where the code was passing.

**Comment by GeneratorX16** [comment]
I tried running the tests from the develop branch locally, and they were failing over there too. Can there be an issue in that ?
I will check and submit again.

**Comment by straker** [comment]
Develop is passing on my end. The way our tests are set up you have to build first before you test, otherwise the changes won't be tested.

**Comment by GeneratorX16** [comment]
My mistake, I was looking at the wrong place 😅. I have debugged the issue now.

**Comment by straker** [comment]
This all looks good, thanks for the updates. I'm trying to finalize feedback from other teams about changing the `html` property, but once I get the all clear I should be able to merge.

**Review by straker — APPROVED** [review]
Sorry this took so long, took a while to get approval from all the other teams involved. They confirmed this change shouldn't break deduplication so can be merged.

**Review by sierra88-en — APPROVED** [review]
__

File: lib/core/utils/dq-element.js (hunk 2 above)
Comment (sierra88-en) [reply lib/core/utils/dq-element.js]: __

Comment (sierra88-en) [reply lib/core/utils/dq-element.js]: H

File: test/core/utils/dq-element.js
Comment (sierra88-en) [reply test/core/utils/dq-element.js:1]: __
